### PR TITLE
ButtonGroup: Adds `list` variant

### LINKED ISSUE #1371: Update Button Group to use list variant
This should allow tooltips trigger `div`'s to be wrapped around buttons within button group (such as in a page header and prevent the following. See list example in  https://www.lightningdesignsystem.com/components/button-groups/

27f603fc-e3d4-11e6-8270-d72288554c08

### PR DESCRIPTION
**Pull request by pulkonet**
Fixes #1371 

Also adds `list` variant for Tooltip component to work well with ButtonGroup.

Screenshot 2019-04-23 at 2 04 58 AM

### Additional description

---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [ ] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [ ] `npm run lint:fix` has been run and linting passes.
* [ ] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] TravisCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

File: components/button-group/index.jsx [hunk 1]
```diff
@@ -102,6 +102,16 @@ const ButtonGroup = (props) => {
 			</fieldset>
 		);
 	}
+	if (props.variant === 'list') {
+		return (
+			<ul
+				className={classNames('slds-button-group-list', props.className)}
+				role="group"
```
Comment (interactivellama) [inline components/button-group/index.jsx]: Group is implicit here and not needed--unlike in base variant

File: components/tooltip/__examples__/button-group.jsx [hunk 2]
```diff
@@ -12,35 +12,44 @@ class Example extends React.Component {
...
-						iconName="down"
-						iconVariant="border-filled"
-						onSelect={(item) => console.log('selected', item)}
-						options={[
-							{ label: 'A Option', value: 'A0' },
-							{ label: 'B Option', value: 'B0' },
-							{ label: 'C Option', value: 'C0' },
-						]}
-					/>
+					<Tooltip
```
Comment (interactivellama) [inline components/tooltip/__examples__/button-group.jsx]: Tooltip isn't working for me here. You are welcome to just remove it if you want. I'm guessing this is a side affect of Dropdown.

Comment (pulkonet) [reply components/tooltip/__examples__/button-group.jsx]: Removing the tooltip is adding double borders ( no overlap ) and is also not adding the `corner radius` to dropdown button.

Comment (pulkonet) [reply components/tooltip/__examples__/button-group.jsx]: I am wrapping it in an `li` tag for now.

File: components/button-group/index.jsx (hunk 1 above)
Comment (pulkonet) [reply components/button-group/index.jsx]: Removed.

File: components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot [hunk 3]
```diff
@@ -1992,41 +1991,45 @@ exports[`DOM snapshots SLDSPopoverTooltip Button Group 1`] = `
...
-      className="slds-dropdown-trigger slds-dropdown-trigger_click slds-button_last"
-      id="options"
-      onClick={[Function]}
-      onFocus={null}
-      onKeyDown={[Function]}
-      onMouseEnter={null}
-      onMouseLeave={null}
+    </li>
+    <li
+      triggerClassName="slds-button_last"
```
Comment (interactivellama) [inline components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]: I'm getting this warning on the example added at http://localhost:9001/?selectedKind=SLDSPopoverTooltip&selectedStory=Button%20Group&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

Screen Shot 2019-05-14 at 1 56 03 AM

The ButtonGroup is expecting `Button` child components which have a `triggerClassName` but, but `li` element tags do not. Please fix this warning.

Comment (pulkonet) [reply components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]:
```if (zeroIndexLength > 0) {
		children = React.Children.map(props.children, (child, index) => {
			let newChild;
			if (index === zeroIndexLength) {
				newChild = React.cloneElement(child, {
					triggerClassName: 'slds-button_last',
				});
			}

			return newChild || child;
		});
	}
```
The error is being generated from the above LOC (present in `button-group`). Adding a type check to prevent the prop `triggerClassName` being added to `li` element will resolve the warning. Shall I go ahead with that ? @interactivellama

Comment (interactivellama) [reply components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]: Yes. I wouldn't recommend variable type checking though. Since `triggerClassName` is an implementation details of `SLDSButton`, you could check for that and only apply the prop if it's an instance of a `SLDSButton`.

Comment (pulkonet) [reply components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]: But we are adding the prop to the instance of `Tooltip` rather than `SLDSButton`. Because the children of the `ButtonGroup` are `Tooltip`s and not `Button`s

Comment (pulkonet) [reply components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]: Do you have any other suggestions in mind?

Comment (interactivellama) [reply components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]: You are exactly right. My mistake. The prop will only work on a `Tooltip` component. So we want to test if the child is a `Tooltip`. 

Here is a displayName test in another component: https://github.com/salesforce/design-system-react/blob/master/components/builder-header/nav.jsx#L33

Comment (pulkonet) [reply components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]: I have added the `displayName` test for the `Tooltip`

File: components/tooltip/__examples__/button-group.jsx (hunk 2 above)
Comment (pulkonet) [reply components/tooltip/__examples__/button-group.jsx]: The error is being generated from here, we can't be adding props to DOM elements in `camelCase`, they need to be in `lowercase`.

So, it would be best if we can either type check the element and filter out `li` or let `Tooltip` be here unless it is creating problems.

@interactivellama

File: components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot (hunk 3 above)
Comment (pulkonet) [reply components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]: It is creating issues elsewhere in the projects wherever `ButtonGroup` is being used. My proposed solution would be to replace `li` with `Tooltip` and let it be there.

Comment (interactivellama) [reply components/tooltip/__docs__/__snapshots__/storybook-stories.storyshot]: If you are referring to PageHeader, https://travis-ci.org/salesforce/design-system-react/builds/534113631#L7524 for instance.

Like code, Jest snapshot markup tests are not fixed in stone. Similar to version control such as "git", the snapshots are a way to track history over time (humans are very bad at this!) and know what has changed across time. If there is a bug or a better way to do something, updating the Jest DOM snapshots is a perfectly fine thing to do. 

In summary, it's OK to change the DOM snapshots, but only if it's a better solution. They are there so that you know what is changing and can decide if you really want to update the markup. Please review the changes across the library per the "failed" Jest tests visually and decide if updating the markup is best and report back.

File: components/tooltip/__examples__/button-group.jsx (hunk 2 above)
Comment (interactivellama) [reply components/tooltip/__examples__/button-group.jsx]: A tooltip around a Dropdown is fine. My issue is that the Tooltip wasn't being triggered on mouseover / "not working". I was thinking that if test the displayName === Tooltip then the extra props wouldn't get passed unless Tooltip was present. I might be misunderstanding though.

Comment (pulkonet) [reply components/tooltip/__examples__/button-group.jsx]: I am reverting to a tooltip around the dropdown instead of li. I will also open another issue for the Tooltip not being triggered and we can work on it there.

**Comment by kevinparkerson** [comment]
@davidlygagnon I'm in need of this variant for the page header work I've been doing so if you don't mind I can review tomorrow and merge when it's good to go?

**Comment by davidlygagnon** [comment]
@kevinparkerson Go for it - sounds good to me.

File: components/button-group/index.jsx [hunk 4]
```diff
@@ -102,6 +99,15 @@ const ButtonGroup = (props) => {
 			</fieldset>
 		);
 	}
+
+	if (props.variant === 'list') {
+		return (
+			<ul className={classNames('slds-button-group-list', props.className)}>
+				{children}
```
Comment (kevinparkerson) [inline components/button-group/index.jsx]: Since every child within ButtonGroup of variant `list` is meant to be wrapped in an `li` tag, please make this code do that here. Something like this:
```
{React.Children.map(props.children, (child) => (
	<li>{child}</li>
))}
```

**Comment by kevinparkerson** [comment]
@pulkonet please create an example of the list variant in use. Put it in it's own file called `list-variant.jsx` beneath `__examples__`. Add it to both `__docs__/site-stories.js` and `__docs__/storybook-stories.jsx`. See [the list variant on SLDS](https://www.lightningdesignsystem.com/components/button-groups/#List) for what this example should look like. Be sure to use actual LDSR Button components for the children rather than the manual buttons seen in that example.

File: components/button-group/index.jsx (hunk 4 above)
Comment (pulkonet) [reply components/button-group/index.jsx]: Done.
